### DISCUSSION TITLE
Multi-partition ordering by input, not alphabetically

### DISCUSSION THREAD

**Discussion by pedrovgp** [dagster-io/dagster#17568] [Ideas]
Today, if I have a multi partitioned asset, the folder levels in storage reflect the alphabetical ordering of the partition names. I would like to suggest it to reflect the ordering that is passed to the MultiPartitionsDefinition.

Example:
```python
tickers_partitions_def = StaticPartitionsDefinition(["tic1", "tic2")
run_day_partitions_def = StaticPartitionsDefinition("2023-04-01", "2023-05-01")
# Partitions are ordered by name, when directories are created
multi_partitions_def = MultiPartitionsDefinition(
    partitions_defs={"ticker": tickers_partitions_def, "run_day": run_day_partitions_def}
)

@asset(
    partitions_def=multi_partitions_def,
    )
def quotes(context) -> List[Dict]:
    ticker = context.partition_key.keys_by_dimension["ticker"]
    run_day = context.partition_key.keys_by_dimension["run_day"]
    return {"ticker": ticker, "run_day": run_day}
```
will create the following directory structure in $dagster_home/storage:
2023-04-01/tic1
2023-04-01/tic2
2023-05-01/tic1
2023-05-01/tic2

If I rename the "run_day" partitions definition to "today"
```python
...
partitions_defs={"ticker": tickers_partitions_def, "today": run_day_partitions_def}
...
```
the order changes to
tic1/2023-04-01
tic1/2023-05-01
tic2/2023-04-01
tic2/2023-05-01

But it is weird to play with names (and find meaningful ones) just to keep an ordering of files that makes more sense (for other systems, for example).

I would like to suggest that the order of `partitions_defs` is respected. So:
```python
multi_partitions_def = MultiPartitionsDefinition(
    partitions_defs={"ticker": tickers_partitions_def, "run_day": run_day_partitions_def}
)
```
would have `ticker` as first level, while  
```python
multi_partitions_def = MultiPartitionsDefinition(
    partitions_defs={"run_day": run_day_partitions_def, "ticker": tickers_partitions_def}
)
```
would have `run_day` as first level.

**Comment by 4l1fe** [comment]
I find out that partitions are explicitly sorted by a reason in here https://github.com/dagster-io/dagster/blob/722b18967c1297e08e8b473ef431f94e016a9e44/python_modules/dagster/dagster/_core/definitions/multi_dimensional_partitions.py#L201

```
self._partitions_defs: List[PartitionDimensionDefinition] = sorted(
    [
        PartitionDimensionDefinition(name, partitions_def)
        for name, partitions_def in partitions_defs.items()
    ],
    key=lambda x: x.name,
)
```

It's quite a big issue, coz a partions order and their names have matter to organize data hierarchy as needed.
